### PR TITLE
Split `BoundInterpolatedString` into `BoundInterpolatedString` and `BoundUnconvertedInterpolatedString`

### PR DESCRIPTION
**Pull request by 333fred**
This is intended to be a semantics-preserving refactoring that will make later work easier. When an interpolated string is first encountered, we now bind it as a `BoundUnconvertedInterpolatedString` node, and then when we bind to the natural type we convert it to a `BoundInterpolatedString` expression. Future code to allow it to convert to non-string types will take advantage of this in the conversion step.

### CODE REVIEW

**Comment by 333fred** [comment]
@chsienki for review.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 1]
```diff
@@ -374,6 +374,18 @@ internal BoundExpression BindToNaturalType(BoundExpression expression, BindingDi
                         result = BindObjectCreationForErrorRecovery(expr, diagnostics);
                     }
                     break;
+                case BoundUnconvertedInterpolatedString unconvertedInterpolatedString:
+                    {
+                        // We determine the best method of emitting as a string (either via Concat, Format, or the builder pattern)
+                        // during lowering, and it's not part of the publicly-visible API, unlike conversion to a builder type.
+                        result = new BoundInterpolatedString(
```
Comment (chsienki) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:381]: I'm not sure I follow this comment. If we do building in lowering, why do we need to have this split here at all? Or am I confusing builder pattern and builder type as the same thing?

My assumption is that we'll add code here in  future PR to create some new kind of `BoundBuilderInterpoltedString` if we detect its actually a builder pattern/type one right?

Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:381]: The code won't be here. The code will be when the interpolated string is converted to a non-`String` type.

Comment (chsienki) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:381]: Hmm, I guess I'm missing something. Why do we need the `BoundUnconvertedInterpolatedString` / `BoundInterpolatedString` if we're converting to another type later on?

Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:381]: > Why do we need the BoundUnconvertedInterpolatedString / BoundInterpolatedString if we're converting to another type later on?

We need it because we want to tell the difference between a string that has been converted and one that has not been converted. This is the same approach we take with other target-typed constructs such as switch expressions and ternaries. Happy to discuss over teams in more depth if you want.

File: src/Compilers/CSharp/Portable/BoundTree/BoundExpression.cs [hunk 2]
```diff
@@ -286,6 +287,14 @@ internal partial class BoundBinaryOperator
         }
     }
 
+    internal partial class BoundUnconvertedInterpolatedString
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/BoundExpression.cs:290]: >BoundUnconvertedInterpolatedString [](start = 27, length = 34)

Can we override the method once in BoundInterpolatedStringBase?

File: src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs [hunk 3]
```diff
@@ -299,7 +299,7 @@ private ExpressionSyntax SkipParensAndNullSuppressions(ExpressionSyntax e)
             BoundExpression convertedExpression = ConvertPatternExpression(
                 inputType, patternExpression, expression, out constantValueOpt, hasErrors, diagnostics);
 
-            ConstantValueUtils.CheckLangVersionForConstantValue(expression, diagnostics);
+            ConstantValueUtils.CheckLangVersionForConstantValue(convertedExpression, diagnostics);
```
Comment (chsienki) [inline src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs:302]: This is an unrelated bug?

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 1 above)
Comment (chsienki) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:381]: Ahh, got it, thanks.

**Review by chsienki — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs (hunk 3 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs:302]: > This is an unrelated bug? Or just expressing itself now?

Neither. Previously, `BoundInterpolatedString` was never converted, so `expression` and `convertedExpression` were the same node.

**Comment by AlekseyTs** [comment]
>                 ? Conversion.InterpolatedString : Conversion.NoConversion;

Should these conversions be allowed only from an unconverted form?

---
Refers to: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs:87 in 01de52a. [](commit_id = 01de52a2efdd11c2dbe7a0bcda2d1554c98afc90, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs [hunk 4]
```diff
@@ -72,7 +72,7 @@ internal ConversionsBase WithNullability(bool includeNullability)
 
         protected abstract ConversionsBase CreateInstance(int currentRecursionDepth);
 
-        protected abstract Conversion GetInterpolatedStringConversion(BoundInterpolatedString source, TypeSymbol destination, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo);
+        protected abstract Conversion GetInterpolatedStringConversion(BoundInterpolatedStringBase source, TypeSymbol destination, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs:75]: >BoundInterpolatedStringBase [](start = 70, length = 27)

It feels like this should be BoundUnconvertedInterpolatedString instead.

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs [hunk 5]
```diff
@@ -1056,6 +1056,11 @@ public override BoundNode VisitInterpolatedString(BoundInterpolatedString node)
             return null;
         }
 
+        public override BoundNode VisitInterpolatedString(BoundInterpolatedString node)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:1059]: Are we not expecting to run into an uconverted node? I think we should still override the visit method and perhaps throw.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 6]
```diff
@@ -9362,6 +9362,14 @@ private static bool IsNullabilityMismatch(TypeSymbol type1, TypeSymbol type2)
             return result;
         }
 
+        public override BoundNode? VisitUnconvertedInterpolatedString(BoundUnconvertedInterpolatedString node)
+        {
+            // This is only involved with unbound lambdas or when visiting the source of a converted tuple literal
+            var result = base.VisitInterpolatedStringBase(node);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:9368]: >VisitInterpolatedStringBase [](start = 30, length = 27)

Why not VisitUnconvertedInterpolatedString?

**Comment by AlekseyTs** [comment]
Done with review pass (commit 1)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 6 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:9368]: With your other suggestion, this would now throw, so I'll leave this as is.

---
In reply to: [601832232](https://github.com/dotnet/roslyn/pull/52061#discussion_r601832232) [](ancestors = 601832232)

**Comment by 333fred** [comment]
@AlekseyTs addressed feedback.

---
In reply to: [807471528](https://github.com/dotnet/roslyn/pull/52061#issuecomment-807471528) [](ancestors = 807471528)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 6 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Updated based on our offline discussion.

---
In reply to: [601847314](https://github.com/dotnet/roslyn/pull/52061#discussion_r601847314) [](ancestors = 601847314,601832232)

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 3)
